chamelon/function_compat.ml

```diff
@@ -0,0 +1,75 @@
+open Typedtree
+open Types
+open Compat
+
+type cases_view = {
+  arg_label : Asttypes.arg_label;
+  param : Ident.t;
+  cases : value case list;
+  partial : partial;
+}
+
+let dummy_type_expr = newty2 ~level:0 (mkTvar (Some "a"))
+
+let mk_exp ed =
+  {
+    exp_desc = ed;
+    exp_loc = Location.none;
+    exp_extra = [];
+    exp_type = dummy_type_expr;
+    exp_env = Env.empty;
+    exp_attributes = [];
+  }
+
+let cases_view_to_function { arg_label; param; cases; partial } =
+  match cases with
+  | [ { c_lhs; c_guard = None; c_rhs } ] -> (
+      let param : texp_function_param =
+        {
+          arg_label;
+          pattern = c_lhs;
+          param;
+          partial;
+          param_identifier = texp_function_param_identifier_defaults;
+        }
+      in
+      match view_texp c_rhs.exp_desc with
+      | Texp_function (inner_function, _) ->
+          { inner_function with params = param :: inner_function.params }
+      | _ -> { params = [ param ]; body = Function_body c_rhs })
+  | cases ->
+      {
+        params = [];
+        body =
+          Function_cases
+            {
+              cases;
+              param;
+              partial;
+              function_cases_identifier =
+                texp_function_cases_identifier_defaults;
+            };
+      }
+
+let function_to_cases_view { params; body } =
+  match (params, body) with
+  | param :: params, body ->
+      let c_rhs =
+        match (params, body) with
+        | [], Function_body body -> body
+        | _, _ -> mk_exp (mkTexp_function { params; body })
```

Comment:
I think this will drop the function identifier, and therefore dropping the return sort and other information with `compat.jst.ml`. This function should probably take an optional `texp_function_identifier` as well; which would need to be passed at the different use sites where it is available.

Reply:
Done!

To summarize a conversation that happened offline: I wondered why it's necessary to preserve the sort, given that it's unused in the `Untypeast`-to-`Pprintast` pipeline that produces the next step of the minimization process. @Ekdohibs reminded me that the identifier tracks mode as well, and regardless she thinks it's a good idea to preserve information in minimization in case other steps depend on this (e.g.) sort information.